torchtune/data/_tokenize.py

```diff
@@ -0,0 +1,151 @@
+
+    return dict(
+        prompt_input_ids=prompt_input_ids,
+        input_ids=answer_input_ids,
+    )
+
+def tokenize_preference_sample(
+        tokenizer: Tokenizer,
+        prompt: str,
+        chosen: str ,
+        rejected: str,
+        max_length: int = 512,
+        max_prompt_length: int = 128,
+        truncation_mode: str = "keep_end",
+) -> Dict:
+    batch = {} 
+    if not isinstance(prompt, str):
+        raise ValueError(f"prompt should be an str but got {type(prompt)}")
+    prompt_tokens = tokenizer.encode(prompt, add_bos=False, add_eos=False)
+    prompt_tokens = {"prompt_input_ids":  prompt_tokens}
+
+    if not isinstance(chosen, str):
+        raise ValueError(f"chosen should be an str but got {type(chosen)}")
+    chosen_tokens = build_tokenized_answer(tokenizer, prompt, chosen)
+
+    if not isinstance(rejected, str):
+        raise ValueError(f"rejected should be an str but got {type(rejected)}")
+    rejected_tokens = build_tokenized_answer(tokenizer, prompt, rejected)
+
+    # Last prompt token might get merged by tokenizer and
+    # it should not be included for generation if that happens
+    prompt_len_input_ids = len(prompt_tokens["prompt_input_ids"])
+
+    chosen_prompt_len_input_ids = len(chosen_tokens["prompt_input_ids"])
+    rejected_prompt_len_input_ids = len(rejected_tokens["prompt_input_ids"])
+    prompt_len_input_ids = min(chosen_prompt_len_input_ids, rejected_prompt_len_input_ids)
+
+    for k, v in prompt_tokens.items():
+        prompt_tokens[k] = v[:prompt_len_input_ids]
+
+    # Make sure prompts only have one different token at most an
+    # and length only differs by 1 at most
+    num_diff_tokens = sum(
+        [a != b for a, b in zip(chosen_tokens["prompt_input_ids"], rejected_tokens["prompt_input_ids"])]
+    )
+    num_diff_len = abs(chosen_prompt_len_input_ids - rejected_prompt_len_input_ids)
+    if num_diff_tokens > 1 or num_diff_len > 1:
+        raise ValueError(
+            "Chosen and rejected prompt_input_ids might only differ on the "
+            "last token due to tokenizer merge ops."
+        )
+
+    # add BOS token to head of prompt
+    prompt_tokens["prompt_input_ids"] = [tokenizer.bos_id] + prompt_tokens["prompt_input_ids"]
+    chosen_tokens["prompt_input_ids"] = [tokenizer.bos_id] + chosen_tokens["prompt_input_ids"]
+    rejected_tokens["prompt_input_ids"] = [tokenizer.bos_id] + rejected_tokens["prompt_input_ids"]
+
+    # add EOS token to end of answer
+    chosen_tokens["input_ids"].append(tokenizer.eos_id)
+    rejected_tokens["input_ids"].append(tokenizer.eos_id)
+
+    longer_response_length = max(len(chosen_tokens["input_ids"]), len(rejected_tokens["input_ids"]))
+
+    # if combined sequence is too long, truncate the prompt
+    for answer_tokens in [chosen_tokens, rejected_tokens, prompt_tokens]:
+        if len(answer_tokens["prompt_input_ids"]) + longer_response_length > max_length:
+            if truncation_mode == "keep_start":
```

Comment:
you should use the `torchtune.data._utils.truncate` utility and add this functionality if needed